### ISSUE TITLE
Rule suggestion: `switch-exhaustiveness-check`

### ISSUE THREAD

**Issue by lionel-rowe** [denoland/deno_lint#1160]
Disallow non-exhaustive `switch` statements. Especially useful when expanding an enum/string union/numeric union type to allow new options.

See https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/docs/rules/switch-exhaustiveness-check.md

**Wrong ❌**
```ts
type X = 'A' | 'B'

const x: X = 'A'

switch (x) {
    case 'A':
        break
}
```

**OK ✅**
```ts
switch (x) {
    case 'A':
        break
    case 'B':
        break
}
```

**OK ✅**
```ts
switch (x) {
    case 'A':
        break
    default:
        break
}
```

**Comment by magurotuna** [comment]
I would love this lint rule, but unfortunately this would require access to type information that TypeScript provides but deno_lint cannot access.

**Comment by lionel-rowe** [comment]
For anyone else that stumbles on this issue, I figured out a workaround via adding a dummy `default` clause. The non-exhaustive switch statement fails type checking with `Type 'string' is not assignable to type 'never'.` [Playground](https://www.typescriptlang.org/play?#code/C4TwDgpgBAGlC8UDkBBJUA+yBCSBQAJhAMYA2AhgE7TED2AdgM7BQAeAXLHngPQ9QADACrhoSZpQCW9AObpJjKPVotyjRpJn1yAI1LRgtKKEjJ6EAG4RKSAHQC8jAO6TgxABZQAFKwCUUAG88AEhiNTE0dhDgnWpyAGsQogAzcgBXUmBOC1pJAm8vX05zK0oEAD42XzwAX24+KAB5AGlHFzdPH38gqF6oMMYIpCi+0ahYiAS8UYGxXBGx3omp4JT0zOzc-K9C4strCqra+v4WttcPbz9A6b7Z5EjbxeXE0bWMrKexl+OgA)

```ts
type X = 'A' | 'B'
declare const x: X

// `Type 'string' is not assignable to type 'never'.`
switch (x) {
	case 'A':
		break
	default: void ((): never => x)
}

// OK
switch (x) {
    case 'A':
        break
    case 'B':
        break
	default: void ((): never => x)
}

// OK
switch (x) {
    case 'A':
        break
    default:
        break
}
```
